### ISSUE TITLE
UnicodeDecodeError: 'ascii' codec can't decode byte 0xfe in position 0: ordinal not in range(128)

### ISSUE THREAD

**Issue by yerpj** [g-oikonomou/sensniff#12]
Is it possible that the sensniff script misunderstands the end of a packet (when, for example, the payload itself contains `C1 1F FE 72`) ?

```

Read 2 bytes but not part of a frame

Traceback (most recent call last):

 File "sensniff.py", line 624, in <module>

   raw = in_handler.read_frame()

 File "sensniff.py", line 171, in read_frame

   logger.info("Peripheral: %s%s" % (b.decode(), per_out.decode()))

UnicodeDecodeError: 'ascii' codec can't decode byte 0xfe in position 0: ordinal not in range(128)
```

**Comment by g-oikonomou** [comment]
That's very useful, is this under PY3 or 2.7?

**Comment by yerpj** [comment]
2.7

**Comment by g-oikonomou** [comment]
OK this could be a bug caused by my recent changes to add PY3 support. I'll test this. Thanks!

**Comment by yerpj** [comment]
I hope this could be easily fixed. I encounter this bug quite frequently, and I always have to keep a look at the script to verify it is still running.

**Comment by g-oikonomou** [comment]
If you can figure out a way to reliably reproduce it, it would help a lot.

The logic in the script is: "If we receive bytes and they are not the magic, we will assume it's the peripheral trying to print something".

Now this can happen if the peripheral is indeed trying to print, if we are using the wrong baud rate, or if the host and the peripheral are out of synch. It could also happen because the script currently still supports the legacy framing format (which I'll remove one day).

**Comment by yerpj** [comment]
Tomorrow I can try to inject packets that contains the 'magic sequence' as payload

**Comment by g-oikonomou** [comment]
I forgot to say, the script will try to print peripheral output (== reception of bytes that do not start with the magic). This is the condition that will result in this particular code branch to execute.

My gut feeling is this won't get you anywhere :) But then again, I've never run this thing on windows so perhaps behaviour could have subtle differences.

**Comment by g-oikonomou** [comment]
Additional idea: The script will try to read 5 bytes in a single go (magic + length). I am now wondering whether for whatever reason this read only returns 2 bytes, which will also confuse the script even if those 2 bytes are part of the magic.

**Comment by yerpj** [comment]
Maybe the 'end of a packet' detection could be determined in two ways:
1) new magic sequence incoming (the one that belongs to a new frame)
2) timeout

This way, it will never try to interpret a packet containing only a part of the magic sequence (assuming the magic sequence last bytes arrival are being delayed by the node)

**Comment by yerpj** [comment]
FYI: 
I just tried to emulate a sniffer HW by injecting packets through a USB-serial built in [Espruino](www.espruino.com):
```
var Magic=[0xC1,0x1F,0xFE,0x72];
var Version=0x02;
var Cmd=0x00;
var Length=[0x00,0x00];

function sendHost(payload,len)
{
  Length[1]=len;
  USB.write(Magic.concat(Version).concat(Cmd).concat(Length).concat(payload));
}

var AssociationResponse=[0x23,0xdc,0x09,0x05,0xf2,0x16,0x00,0x3f,0x00,0x32,0x31,0x35,0x35,0x00,0x00,0x45,0x00,0x1a,0x00,0x32,0x31,0x35,0x35,0x02,0xff,0xfe,0x00,0xcb,0x80];

pinMode(LED2,'output');
setInterval(function(){
  sendHost(AssociationResponse,AssociationResponse.length);
  digitalWrite(LED2,!digitalRead(LED2));
},500);
```

This script sends every 500 milliseconds the same packets(Association Response) . Sensniff.py already received and forwarded almost 1000 messages. 

Before that, I forgot to send the correct [length], and the same issue raised up everytime on the first packet.

**Comment by yerpj** [comment]
I made a few more tests:
1) When I inject a packet with correct "sensniff" header but with wrong IEEE802.15.4 syntax,nothing odd happens. --> correct behavior
2) When I inject a packet with a correct IEEE802.15.4 packet but with wrong "sensniff" `length`header (real length -1), sensniff tells me `Read 1 bytes but not part of a frame`, and forward anyway the packet to Wireshark, which detects that the packet is malformed. --> correct behavior
3) When I inject a packet with a correct IEEE802.15.4 packet but with wrong "sensniff" `length`header (real length +1), sensniff tells me (after the second packet) `Read correct header not followes by a frame \n Expected 30 bytes, got 29` and don't send anythint to Wireshark. --> correct behaviour.
4) When the injected packet has a payload containing the magic sequence, sensniff acts totally as expected. --> correct behavior
5) When I concat 2 sensniff packets and send them to sensniff, everything works as expected.--> correct behavior
6) I retried to corrupt the sensniff `length`field with different values. (Regular length should be 29)
- length=1:
```Traceback (most recent call last):
  File "sensniff.py", line 624, in <module>
    raw = in_handler.read_frame()
  File "sensniff.py", line 183, in read_frame
    logger.info("Peripheral: %s%s" % (b.decode(), per_out.decode()))
UnicodeDecodeError: 'ascii' codec can't decode byte 0xdc in position 0: ordinal not in range(128)
``` 
-length=28:
```
Read 1 bytes but not part of a frame
```
-length=27:
```
Read 2 bytes but not part of a frame
```
-length=26:
```
Read 3 bytes but not part of a frame
```
-length=25:
```
Read 4 bytes but not part of a frame
```
-length=24:
```
Traceback (most recent call last):
  File "sensniff.py", line 624, in <module>
    raw = in_handler.read_frame()
  File "sensniff.py", line 183, in read_frame
    logger.info("Peripheral: %s%s" % (b.decode(), per_out.decode()))
UnicodeDecodeError: 'ascii' codec can't decode byte 0xff in position 0: ordinal not in range(128)
```
It seems that something non linear happens when the length field is corrupted. Anyway, each packet sent to sensniff should be valid. In my case, I suspect my sniffer hardware to mis-format the ieee802.15.4 packets before sending them to Host.

**Comment by yerpj** [comment]
I made another experiment: I sniffed traffic both with a node+sensniff+wireshark, as well as with a custom sniffer (console based). When sensniff returned the error `UnicodeDecodeError: 'ascii' codec can't decode byte 0xfe in position 0: ordinal not in range(128)` I analyzed the last packet sent to wireshark, and I found its equivalent on my custom console-based sniffer. Then I had a look at the following packets, thinking that one of those would have something odd (such as wrong length or anything else). Result: None of the following packets were malformed. I am thinking again that the serial handler of sensniff may be the cause of the crash.

**Comment by yerpj** [comment]
I suspect the serial timeout to be a little bit too short (current timeout == 0.1 sec) . I am performing tests with timeout=0.5

**Comment by yerpj** [comment]
I apologize for the number of comments on this issue. This may be the last one, because I think I found the problem.
I injected packets but this time I artificially sliced them in two chunks and I sent those two chunks sequentially, with a time `T_delay` between them. For `T_delay` lower than the Serial.timeout, everything happens as expected. But for `T_delay` greater than Serial.timeout, the script goes in error. 

With Contiki, and its very clever implementation (maybe thanks to protothreads too), the serial output can't be interrupted, or at least not for a time greater than 100ms. This explains why this issue was never crossed. With my custom sniffer, I am almost certain that a task with greater priority than the console task can interrupt the latter for times that could be greater than 100 or 200ms (I will have to cross-check the task priorities in my system ;-)  )

This is why I encountered problems with my setup. I increased Serial.timeout up to 1 second, and now everything works as expected. I already sniffed more than 4'000 packets and everything works like a charm. AFAIK this relaxed timeout doesn't add any drawback to the script, except maybe a small delay before printing the packets to wireshark but if the packet is still incoming from serial, in any way, sensniff has to wait on it before making anything else.

**Comment by g-oikonomou** [comment]
Thanks for all the input.

I see that in the OP the error was in line 171 whereas in subsequent posts it appears in L183. Can you confirm that it's actually the same code block that fails in all cases?

```
        if b[0:4] not in (self.__sensniff_magic, self.__sensniff_magic_legacy):
            # Peripheral UART output - print it
            per_out = self.port.readline().rstrip()
            logger.info("Peripheral: %s%s" % (b.decode(), per_out.decode()))
```

The error happens when sensniff thinks the peripheral is trying to print readable output over the serial line, when in reality your device is sending a frame. This happens because earlier in the script, `size` was less than 5. The reason for this to occur during streaming a magic could be due to the timeout. Which makes me think: What baud rate are you using on your device.

This is an ugly error obviously. What should really happen is the script should log a warning and continue working. There is a try / except missing around that code region :)

**Comment by yerpj** [comment]
>Can you confirm that it's actually the same code block that fails in all cases? 

I confirm, this is the same failure.

**Comment by g-oikonomou** [comment]
Ah OK that explains it.

I should add that this particular condition probably happens when your task is interrupted while sending out the magic.

If interrupted while sending out the actual frame, you are probably looking at the branch near L189 to execute: `if len(b) != size:`

**Comment by yerpj** [comment]
that's correct.

**Comment by g-oikonomou** [comment]
Sooooo

What's a fail-safe value for the timeout? ;)

**Comment by yerpj** [comment]
in my case I am using 1 second, as this has no consequence on the global behavior

**Comment by g-oikonomou** [comment]
Muppetry: I posted this under the PR instead of here.

Joking aside, this will be tricky to handle. The script reads in 5-byte chunks, expecting to see the magic + version bytes. When this condition occurs, those reads will only become aligned to the start of a frame when the peripheral has stayed silent for more than the timeout. This may end up taking quite a long time, with many frames discarded in between.

Gets even worse. Once the script reads the correct header, it will then try to read out the entire remainder of the frame in one go (near line 226 `b = self.port.read(length)`). This has to happen within the timeout. For the current timeout of 100ms at the default of 460800, the theoretical maximum is 46080 bits = 5760 bytes. In the future I want to be able to support long 802.15.4g frames (up to 2KB), so this condition may become more frequent.

However, a good start is to get the script to handle this condition without blowing up and I have pushed branch `fix-unicode-error`. My testbed is not set up at the moment, so could I ask you to revert timeout to 0.1 and try with this branch?

**Comment by g-oikonomou** [comment]
I should add, a timeout value of 1 sec will make interactive mode less responsive (up to 1sec between user input and the script reacting to it.

**Comment by yerpj** [comment]
Yes I can proceed to some tests on the new branch.

Serial parsers are  tricky when packets don't have any footer. In this case, if you expect to keep a timeout of 100ms, maybe it should be written in the README that a packet HAS TO be comprised within a time windows of 100ms

**Comment by g-oikonomou** [comment]
Yep, I too was thinking this needs documented (and have thought about the footer thing too).

**Comment by g-oikonomou** [comment]
Fixed in #15
